File: zundoko/zundoko.pl
sub zundoko {
    my @zundoko = ("ズン", "ドコ");
    my @target = ("ズン", "ズン", "ズン", "ズン", "ドコ");
    my $target_str = join('', @target);
    my @history = ();

    while (join('', @history) ne $target_str) {
        my $next = $zundoko[rand(@zundoko)];
        print $next;

        shift @history if ($#history >= $#target);
        push @history, $next;
    }
    print "キヨシ！\n";
}

for (my $i = 0; $i < 10000; $i++) {
    zundoko();
}

File: zundoko/zundoko.js
// 配列からランダムに値を取り出す
function sample(arr) {
  const n = Math.floor(Math.random() * arr.length);
  return arr[n];
}

function zundoko() {
  const zundoko = ["ズン", "ドコ"];
  const target = ["ズン", "ズン", "ズン", "ズン", "ドコ"];
  const target_str = target.join('');
  const history = [];

  while (history.join('') != target_str) {
    let next = sample(zundoko);
    process.stdout.write(next);

    if (history.length >= target.length) {
      history.shift();
    }
    history.push(next);
  }
  console.log("キヨシ！");
}

for (let i = 0; i < 10000; i++) {
  zundoko();
}
